diff --git a/src/main/Qafoo/ChangeTrack/Analyzer/ChangeRecorder.php b/src/main/Qafoo/ChangeTrack/Analyzer/ChangeRecorder.php
index 717b40e..73062ad 100644
--- a/src/main/Qafoo/ChangeTrack/Analyzer/ChangeRecorder.php
+++ b/src/main/Qafoo/ChangeTrack/Analyzer/ChangeRecorder.php
@@ -19,8 +19,6 @@ class ChangeRecorder
     public function recordChange(\ReflectionClass $class, \ReflectionMethod $method)
     {
         if (!isset($this->changes[$class->getName()])) {
-            $this->changes[$class->getName()] = array();
-        }
         $this->changes[$class->getName()][$method->getName()] = 1;
     }
 
